book/src/rfc/rfc-0-generic-parameters.md

```diff
@@ -0,0 +1,359 @@
+    fn_call(ii);
+}
+```
+
+To allow generic functions in for-loop, we will need to take care of unrolling the loop and instantiating the function with the concrete value of the loop variable. This is not in the scope of this RFC.
+
+*Forbid operations on symbolic value of arguments*
+```rust
+// disallow arithmetic operations on the symbolic value of the function arguments,
+// such as NN * 2 in this case.
+// because it is challenging to resolve the value of NN.
+fn last(arr: [Field; NN * 2]) -> Field {
+    return arr[NN - 1];
+}
+```
+
+*Defer type checks*
+Anything involves the generic parameters should be deferred to MAST phase. We need to defer the type checks for array with generic size.
+
+In MAST phase, the values of generic parameters can be resolved, so the symbolic values can be evaluated. Thus, all the types with generic parameters can be type checked, as the array sizes become concrete values. 
+
+### MAST
+After the TAST phase, the MAST phase can resolve the generic values from the observed arguments by propagate the constant values through the main function AST.
+
+### Resolving algorithm
+The algorithm will need to handle the following two categories:
+- Resolving from constant argument
+- resolving from array argument
+
+Example of resolving constant arguments:
+```rust
+// constant generic
+// - the value of LEN can be resolved from an observed constant value propagated
+// - store the value in the function body scope
+fn gen_arr(const LEN: Field) -> [Field; LEN] {
+    let arr = [Field; LEN];
+    return arr;
+}
+```
+
+
+Example of resolving from array arguments:
+```rust
+// First, here is a simple case.
+// - the LEN can be resolved from the array size of argument `arr` 
+// - store the value of N in the context
+fn last(arr: [Field; LEN]) -> Field {
+    return arr[LEN - 1];
+}
+```
+
+```rust
+// Then, here is a more challenging case that would require techniques like SMT to solve the unique values of the generic parameters.
+// Even LEN * 2 looks obvious to solve, solving it may need something like (computer algebra system) CAS in rust.
+// It is easy to evaluate the Symbolic values using the solved generic value. But the way around is difficult.
+fn last(arr: [Field; LEN * 2]) -> Field {
+    return arr[LEN - 1];
+}
+```
+
+In this RFC, we want to just enforce the syntax to be sufficient to support the simple cases. That is to disallow arithemtic operations among generic parameters for function arguments. 
+
+
+To recap, here is the pseudo code for the resolving algorithm for function
+```rust
+// at function caller scope
+// collect observed arguments
+//  at function callee scope
+//    for each argument
+//      if the argument involves generic parameters
+//        resolve the generic values from observed arguments
+//    for each statement
+//      compute the type with the resolved generic values
+//      type checks
+//  return type
+```
```

Comment:
Maybe a list of text instead of a codeblock of indented comments would be better :p